File: 1_dfs_bfs/이준우/타겟 넘버.kt
class Solution {
    fun solution(numbers: IntArray, target: Int): Int {

        // 현재 인덱스, 현재 값
        fun dfs(idx: Int, cur: Int): Int {
            // 마지막 노드까지 탐색을 완료한 이후
            if(idx >= numbers.size) {
                // target이랑 동일한 경우 1을 반환, 그렇지 않으면 0을 반환
                // 동일한 값이 하나 존재한다는 의미로 1을 반환 즉, 카운팅의 의미
                return if(cur == target) 1 else 0
            }
            
            // 현재 값을 더한 경우, 뺀 경우를 나눠서 탐색
            return dfs(idx + 1, cur + numbers[idx]) + dfs(idx + 1, cur - numbers[idx])
        }
        
        return dfs(0, 0)
    }
}

File: 1_dfs_bfs/최영재/타겟넘버.py
def solution(numbers, target):
    return dfs(
        numbers=numbers,
        target=target,
        current_index=0,
        current_sum=0,
    )


def dfs(numbers, target, current_index, current_sum):
    is_decided = current_index == len(numbers)

    if is_decided:
        reached_target = current_sum == target
        if reached_target:
            return 1
        return 0

    current_number = numbers[current_index]

    ways_when_plus = dfs(
        numbers=numbers,
        target=target,
        current_index=current_index + 1,
        current_sum=current_sum + current_number,
    )

    ways_when_minus = dfs(
        numbers=numbers,
        target=target,
        current_index=current_index + 1,
        current_sum=current_sum - current_number,
    )

    return ways_when_plus + ways_when_minus


if __name__ == "__main__":
    print(solution([1, 1, 1, 1, 1], 3))  # 5
    print(solution([4, 1, 2, 1], 4))     # 2
